Crash in WebCore::FragmentedSharedBuffer::takeData

jyavenard opened the pull request:
#### 59ff57feba67aff9725abc7f3cdc7cce5150cd6a
<pre>
Crash in WebCore::FragmentedSharedBuffer::takeData
<a href="https://bugs.webkit.org/show_bug.cgi?id=296427">https://bugs.webkit.org/show_bug.cgi?id=296427</a>
<a href="https://rdar.apple.com/155578324">rdar://155578324</a>

Reviewed by Jer Noble.

Attempting to create a blob from fetched data greater than 4GiB.
Creating a Blob from fetch request was done by concatenating all the content
received into a Vector. As a WTF::Vector can&apos;t contain more than `std::numeric_limits&lt;unsigned&gt;::max()`
elements, a crash would crash.

To get around this, if the resulting blob is going to be more than 512MiB
we generate a fragmented Blob instead made of each of the partial data
received by the fetch command.

TODO: How to test without storing more than 4GiB data on a server?

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::takeAsBlob):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::FragmentedSharedBuffer::segments const): Add convenience method

Canonical link: <a href="https://commits.webkit.org/297885@main">https://commits.webkit.org/297885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/894a63b6d33aaf247bff04ba1c8d60969081e37d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33007 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119475 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63934 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41565 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86207 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26883 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101890 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66533 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20016 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63229 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96257 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122692 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40345 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30097 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95054 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40730 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94800 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24189 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39946 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17762 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36481 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40231 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45730 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39872 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41609 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -224,6 +224,7 @@ class FragmentedSharedBuffer : public ThreadSafeRefCounted<FragmentedSharedBuffe
     DataSegmentVector::const_iterator begin() const LIFETIME_BOUND { return m_segments.begin(); }
     DataSegmentVector::const_iterator end() const LIFETIME_BOUND { return m_segments.end(); }
     bool hasOneSegment() const { return m_segments.size() == 1; }
+    size_t segments() const { return m_segments.size(); }
```
Nit: Can we rename this `segmentCount()` or something?

jernoble reviewed (approved):
r=me with nit

jyavenard committed:
Crash in WebCore::FragmentedSharedBuffer::takeData
https://bugs.webkit.org/show_bug.cgi?id=296427
rdar://155578324

Reviewed by Jer Noble.

Attempting to create a blob from fetched data greater than 4GiB.
Creating a Blob from fetch request was done by concatenating all the content
received into a Vector. As a WTF::Vector can't contain more than `std::numeric_limits<unsigned>::max()`
elements, a crash would crash.

To get around this, if the resulting blob is going to be more than 512MiB
we generate a fragmented Blob instead made of each of the partial data
received by the fetch command.

TODO: How to test without storing more than 4GiB data on a server?

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::takeAsBlob):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::FragmentedSharedBuffer::segments const): Add convenience method

Canonical link: https://commits.webkit.org/297885@main
